services/localfile/server/localfile.go

```diff
@@ -774,6 +796,50 @@ func (s *server) Symlink(ctx context.Context, req *pb.SymlinkRequest) (*emptypb.
 	return &emptypb.Empty{}, nil
 }
 
+func (s *server) Mkdir(ctx context.Context, req *pb.MkdirRequest) (_ *emptypb.Empty, retErr error) {
+	logger := logr.FromContextOrDiscard(ctx)
+	recorder := metrics.RecorderFromContextOrNoop(ctx)
+
+	if req.GetDirDescription() == nil {
+		recorder.CounterOrLog(ctx, localfileMkdirFailureCounter, 1, attribute.String("reason", "missing_dst"))
+		return nil, status.Error(codes.InvalidArgument, "destination must be filled in")
+	}
+	d := req.GetDirDescription()
+
+	a := d.GetAttrs()
+	if a == nil {
+		recorder.CounterOrLog(ctx, localfileMkdirFailureCounter, 1, attribute.String("reason", "missing_attrs"))
+		return nil, status.Errorf(codes.InvalidArgument, "must send attrs")
+	}
+	filename := a.Filename
+	logger.Info("create directory", filename)
+	dir, _, err := setupOutputDir(a)
+	if err != nil {
+		recorder.CounterOrLog(ctx, localfileMkdirFailureCounter, 1, attribute.String("reason", "setup_output_err"))
+		return nil, err
+	}
+
+	cleanup := func() {
+		if retErr != nil {
+			os.Remove(dir)
+		}
+	}
+	defer cleanup()
+
+	// Do one final check (though racy) to see if the file exists.
+	_, err = os.Stat(filename)
+	if err == nil {
+		return nil, status.Errorf(codes.Internal, "directory %s already exists ", filename)
+	}
+
+	// Rename tmp directory to real destination.
+	if err := os.Rename(dir, filename); err != nil {
+		return nil, status.Errorf(codes.Internal, "error renaming %s -> %s - %v", dir, filename, err)
+	}
```

Comment:
I'm not clear on the purpose of the create + rename dance for directories. What do we gain by doing it as opposed to just creating directory in the destination right away?

I know there is no atomic way to mkdir+set owner, but the probability of first succeeding and 2nd failing is very low. We are not writing a database here, so maybe we can ignore that in the name of simplicity?

Reply:
Yes, that's the main consideration by create a tmp (setting ownership) + rename to make then atomic. But your suggestion also works, we can also directly create a directory and once some wrong happen with ownership or mode setting, os.Remove(dirName). @sfc-gh-srhodes looking forward to your opinion.

Reply:
Let's do the simpler in-place implementation for now. We can always change it to a more complex implementation later on if we discover issues.